### ISSUE TITLE
as_draws() does not work correctly on mcmc.list objects even though it is supported by more specific as_draws_...() functions

### ISSUE THREAD

**Issue by mjskay** [stan-dev/posterior#115]
It appears that `as_draws()` in selecting an appropriate target draws format "XXX" for an object bypasses S3 dispatch on the chosen `as_draws_XXX()` function, which has the slightly strange (from a user perspective) result that `as_draws(x)` may not be the same as `as_draws_XXX(x)` even if "XXX" is the format chosen by `as_draws()`. 

For example, here is the (incorrect) translation of an `"mcmc.list"` object into a draws object produced by `as_draws()`:

```r
data(line, package = "coda")
as_draws(line)
```
```
# A draws_list: 600 iterations, 1 chains, and 2 variables

[chain = 1]
$line1
 [1] 7.2 3.0 3.7 3.3 3.7 3.6 2.7 3.0 3.5 2.1

$line2
 [1] 2.1 2.6 3.7 2.9 3.1 3.0 2.4 3.3 2.7 3.0

# ... with 590 more iterations
```

And the correct output given by `as_draws_list()`:

```r
as_draws_list(line)
```
```
# A draws_list: 200 iterations, 2 chains, and 3 variables

[chain = 1]
$alpha
 [1] 7.2 3.0 3.7 3.3 3.7 3.6 2.7 3.0 3.5 2.1

$beta
 [1] -1.57  1.50  0.63  1.18  0.49  0.21  0.88  1.09  1.07  1.48

$sigma
 [1] 11.23  4.89  1.40  0.66  1.36  1.04  1.29  0.46  0.63  0.91


[chain = 2]
$alpha
 [1] 2.1 2.6 3.7 2.9 3.1 3.0 2.4 3.3 2.7 3.0

$beta
 [1] 0.95 0.53 0.52 0.98 0.94 0.66 0.52 0.94 1.06 0.83

$sigma
 [1] 2.85 1.71 0.79 0.95 0.90 0.35 1.09 0.97 0.69 0.64

# ... with 190 more iterations
```

The problem appears to stem from this location:

https://github.com/stan-dev/posterior/blob/d0845b3f606a09c51da73e03a6a3ebe2b3c1389d/R/as_draws.R#L47-L52

which uses the internal `.as_draws_XXX()` functions instead of `as_draws_XXX()` functions. This makes sense, since otherwise you'd end up with infinite recursion if the object passed to `as_draws_XXX()` ended up in `as_draws_XXX.default()` due to calls back into `as_draws()`, like this one:

https://github.com/stan-dev/posterior/blob/d0845b3f606a09c51da73e03a6a3ebe2b3c1389d/R/as_draws_list.R#L26-L29

However, it does lead to these weird mismatches that are odd from the user side.

I can see two solutions off the top of my head:

1. Use some introspection to check for a non-default implementation of `as_draws_XXX()` and call to that instead of `.as_draws_XXX()` if one is found. Shouldn't be too annoying, I had to do something like this in tidybayes.
2. Ensure there's an implementation of `as_draws()` for every class supported by one of the `as_draws_XXX()` functions (pro: less magic, con: extra boilerplate).

Thoughts?

**Comment by paul-buerkner** [comment]
I would prefer (1) if there is a safe way of doing that.

**Comment by Ozan147** [comment]
I've run into some other small transformation issues while looking into this. A couple example:
```r
## as_draws_df() does not correctly transform multi-dimensional rvar objects
> as_draws_df(rvar(array(rnorm(10 * 4 * 3), dim = c(10, 4, 3))))
# # A draws_df: 0 iterations, 0 chains, and 0 variables
# data frame with 0 columns and 0 rows
# # ... hidden reserved variables {'.chain', '.iteration', '.draw'}

## Error below cen be a bit misleading because the real issue is object dimension, not class
> as_draws_df(array(1:12, c(2, 3, 2, 1)))
# Error: Don't know how to transform an object of class 'array' to any supported draws format.
```

I believe the entire transformation logic can be restructured without having separate `as_draws_*()` and `.as_draws_*()`, which already call each other. We could condense class transformation methods substantially by moving the existing `.as_draws_*()` logic into default methods. There might be some room for performance improvements as well with the vectorization of some operations.

**Comment by paul-buerkner** [comment]
Thank you for looking into this!

**Comment by jgabry** [comment]
Indeed, thanks! 

> I believe the entire transformation logic can be restructured without having separate `as_draws_*()` and `.as_draws_*()`, which already call each other. We could condense class transformation methods substantially by moving the existing `.as_draws_*()` logic into default methods. There might be some room for performance improvements as well with the vectorization of some operations.

That would be great.

**Comment by paul-buerkner** [comment]
Fixed via #171
